src/main/java/hexgui/gui/HexGui.java

```diff
@@ -0,0 +1,2393 @@
+    Runnable callback = null;
+    if (cb != null) callback = new GuiRunnable(cb);
+    sendCommand(cmd + "\n", callback);
+  }
+
+  /**
+   * HtpShell Callback. By the name of the command it choose the proper callback function. Arguments
+   * are passed as given.
+   */
+  public void commandEntered(String cmd) {
+    sendCommand(cmd, null);
+  }
+
+  // ----------------------------------------------------------------------
+
+  private boolean commandNeedsToLockGUI(String cmd) {
+    if ((cmd.length() > 7 && cmd.substring(0, 7).equals("genmove"))
+        || (cmd.length() > 15 && cmd.substring(0, 15).equals("dfs-solve-state"))
+        || (cmd.length() > 16 && cmd.substring(0, 16).equals("dfpn-solve-state"))
+        || (cmd.length() > 23 && cmd.substring(0, 23).equals("dfs-solver-find-winning"))
+        || (cmd.length() > 24 && cmd.substring(0, 24).equals("dfpn-solver-find-winning")))
+      return true;
+    return false;
+  }
+
+  private void lockGUI() {
+    m_locked = true;
+    m_toolbar.lockToolbar();
+  }
+
+  private void unlockGUI() {
+    m_toolbar.unlockToolbar(m_current, this);
+    m_locked = false;
+  }
+
+  /** A (command, callback) pair. */
+  private class HtpCommand {
+    public HtpCommand() {}
+
+    public HtpCommand(String cmd, Runnable callback) {
+      this.str = cmd;
+      this.callback = callback;
+    }
+
+    public String str;
+    public Runnable callback;
+  }
+
+  /** Waits for commands to be added to the queue, then processes each in turn. */
+  private class CommandHandler implements Runnable {
+
+    public CommandHandler(Component parent, ArrayBlockingQueue<HtpCommand> queue) {
+      m_parent = parent;
+      m_queue = queue;
+    }
+
+    public void run() {
+      while (true) {
+        HtpCommand cmd = null;
+        try {
+          // block until queue contains an element
+          cmd = m_queue.take();
+        } catch (InterruptedException e) {
+          System.out.println("INTERRUPTED! HUH?");
+        }
+
+        if (m_white != null && m_white.connected()) {
+          if (commandNeedsToLockGUI(cmd.str)) lockGUI();
+
+          try {
+            m_white.sendCommand(cmd.str);
+            if (cmd.callback != null) {
+              cmd.callback.run();
+            }
+          } catch (HtpError e) {
+            System.out.println("Caught error '" + e.getMessage() + "'");
+            ShowError.msg(m_parent, e.getMessage());
+          }
+
+          if (commandNeedsToLockGUI(cmd.str)) unlockGUI();
+        } else {
+          System.out.println("Not sending to disconnected: '" + cmd.str.trim() + "'");
+        }
+      }
+    }
+
+    Component m_parent;
+    ArrayBlockingQueue<HtpCommand> m_queue;
+  }
+
+  private void sendCommand(String cmd, Runnable callback) {
+    if (m_white == null) return;
+
+    try {
+      System.out.println("sendCommand: '" + cmd.trim() + "'");
+      m_htp_queue.put(new HtpCommand(cmd, callback));
+    } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)